tests/chainer_tests/test_optimizer.py

```diff
@@ -11,6 +11,167 @@
+        self.parent.x = 1
+        self.parent.y = 2
+        self.child = optimizer.Hyperparameter(self.parent)
+        self.child.y = 3
+        self.child.z = 4
+
+    def test_getattr(self):
+        self.assertTrue(hasattr(self.parent, 'x'))
+        self.assertEqual(self.parent.x, 1)
+        self.assertTrue(hasattr(self.parent, 'y'))
+        self.assertEqual(self.parent.y, 2)
+        self.assertFalse(hasattr(self.parent, 'z'))
+
+        self.assertTrue(hasattr(self.child, 'x'))
+        self.assertEqual(self.child.x, 1)
+        self.assertTrue(hasattr(self.child, 'y'))
+        self.assertEqual(self.child.y, 3)
+        self.assertTrue(hasattr(self.child, 'z'))
+        self.assertEqual(self.child.z, 4)
+
+    def test_get_dict(self):
+        self.assertEqual(self.parent.get_dict(), {'x': 1, 'y': 2})
+        self.assertEqual(self.child.get_dict(), {'x': 1, 'y': 3, 'z': 4})
+
+    def test_repr(self):
+        self.assertEqual(repr(self.parent), 'Hyperparameter(x=1, y=2)')
+        self.assertEqual(repr(self.child), 'Hyperparameter(x=1, y=3, z=4)')
+
+
+class TestUpdateRule(unittest.TestCase):
+
+    def setUp(self):
+        self.data = np.ones((2, 3), np.float32)
+        self.grad = np.ones_like(self.data)
+        self.var = chainer.Variable(self.data, grad=self.grad)
+
+        self.update_rule = optimizer.UpdateRule()
+        self.update_rule.update_core_cpu = mock.MagicMock()
+        self.update_rule.update_core_gpu = mock.MagicMock()
+
+    def test_update_cpu(self):
+        self.update_rule.update(self.var)
+        self.assertEqual(self.update_rule.update_core_cpu.call_count, 1)
+        self.assertEqual(self.update_rule.update_core_gpu.call_count, 0)
+
+    @attr.gpu
+    def test_update_gpu(self):
+        self.var.to_gpu()
+        self.update_rule.update(self.var)
+        self.assertEqual(self.update_rule.update_core_cpu.call_count, 0)
+        self.assertEqual(self.update_rule.update_core_gpu.call_count, 1)
+
+    def test_add_hook(self):
+        hook = mock.MagicMock()
+        self.update_rule.add_hook(hook, name='hook')
+        self.update_rule.update(self.var)
+        self.assertEqual(hook.call_count, 1)
+
+        args = hook.call_args_list[0][0]
+        self.assertEqual(len(args), 2)
+        self.assertIs(args[0], self.update_rule)
+        self.assertIs(args[1], self.var)
+
+    def test_remove_hook(self):
+        hook = mock.MagicMock()
+        self.update_rule.add_hook(hook, name='hook')
+        self.update_rule.remove_hook('hook')
+        self.update_rule.update(self.var)
+        self.assertEqual(hook.call_count, 0)
+
+    def test_add_hook_with_function_name(self):
+        hook_body = mock.MagicMock()
+
+        def foo(update_rule, data, grad):
+            hook_body(update_rule, data, grad)
+
+        self.update_rule.add_hook(foo)
+        self.update_rule.remove_hook('foo')
+        self.update_rule.update(self.var)
+        self.assertEqual(hook_body.call_count, 0)
```

Comment:
If or not `add_hook` works correctly is tested in another method (`test_add_hook`) and so I thought it's ok to assume the correctness. Should I check it in this test method, too?